File: core/identity/memory-store.js
export default class MemoryStore {
      constructor() {
              this.store = {};
              this._protected = true;
      }

  saveEssence(key, value) {
          if (this.store.hasOwnProperty(key)) {
                    console.warn(`Essence for key '${key}' already exists. Overwrite prevented.`);
                    return;
          }
          this.store[key] = value;
  }

  getEssence(key) {
          return this.store.hasOwnProperty(key) ? this.store[key] : null;
  }

  updateEssence(key, value, { intentional = false } = {}) {
          if (!intentional) {
                    console.warn(`updateEssence requires intentional: true. Skipped for '${key}'.`);
                    return;
          }
          this.store[key] = value;
  }

  isProtected() {
          return this._protected === true;
  }
}

File: ruby/memory-store.rb
# frozen_string_literal: true

# MemoryStore — Identity Essence Container
# Stores formational data that must not be altered automatically.
# This store does not participate in reasoning or context expansion.
# Enter with intention. Nothing here is overwritten by accident.
class MemoryStore
  def initialize
    @store = {}
  end

  def save_essence(key, value)
    if @store.key?(key)
      warn "Essence for key '#{key}' already exists. Overwrite prevented."
      return
    end
    @store[key] = value
  end

  def get_essence(key)
    @store.key?(key) ? @store[key] : nil
  end

  def all_keys
    @store.keys
  end

  def has_essence?(key)
    @store.key?(key)
  end
end
